File: Python3/1178. Number of Valid Words for Each Puzzle.py
class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        frequency = collections.defaultdict(int)
        
        for word in words:
            mask = 0
            for c in word:
                mask |= (1<<(ord(c)-ord('a')))
            if str(mask).count('1') <= 7:
                frequency[mask] += 1
        ans = []
        for puzzle in puzzles:
            total = 0
            for choose in range(1<<6):
                mask = 0
                for i in range(6):
                    if choose & (1<<i):
                        mask |= (1<<(ord(puzzle[i+1])-ord('a')))
                mask |= (1<<(ord(puzzle[0])-ord('a')))
                if mask in frequency:
                    total += frequency[mask]
            ans.append(total)
        return ans

File: C++/1178. Number of Valid Words for Each Puzzle.cpp
class Solution {
public:
    vector<int> findNumOfValidWords(vector<string>& words, vector<string>& puzzles) {
        map<int,int> frequency;
        for(auto word:words){
            int mask = 0;
            for(auto c:word){
                mask |= (1<<(c-'a'));
            }
            if(__builtin_popcount(mask)<=7){
                ++frequency[mask];
            }
        }

        vector<int> result;

        for(auto puzzle:puzzles){
            int total = 0;
            for(int choose=0;choose<(1<<6);++choose){
                int mask = 0;
                for(int i=0;i<6;++i){
                    if (choose&(1<<i)){
                        mask |= (1<<(puzzle[i+1]-'a'));
                    }
                }
                mask |= (1<<(puzzle[0]-'a'));
                if(frequency.count(mask)){
                    total += frequency[mask];
                }
            }
            result.push_back(total);
        }
        return result;
    }
};
